### DISCUSSION TITLE
[help] `tar_eval()` cannot track targets nested in list

### DISCUSSION THREAD

**Discussion by psychelzh** [ropensci/tarchetypes#153] [Help] [answered]
### Help

- [X] I understand and agree to <https://books.ropensci.org/targets/help.html>.

### Description

Please ask your question and provide a reprex if troubleshooting.

For example:

```r
# with issues
targets::tar_dir({
  targets::tar_script({
    library(targets)
    list(
      tar_target(x, 1),
      tar_target(y, 2),
      tarchetypes::tar_eval(
        tar_target(name, list(val)),
        values = tibble::tribble(
          ~name, ~val,
          rlang::sym("t"), rlang::syms(c("x", "y"))
        )
      )
    )
  })
  targets::tar_visnetwork()
})
```

The dependency between `t` and `x`, `y` does not work.

![image](https://github.com/ropensci/tarchetypes/assets/13306020/db757d73-11ef-4baa-81e5-b6f03c7d1376)

For now  I can use `tar_map()`:

```r
# workaround?
targets::tar_dir({
  targets::tar_script({
    library(targets)
    list(
      tar_target(x, 1),
      tar_target(y, 2),
      tarchetypes::tar_map(
        values = tibble::tribble(
          ~name, ~val,
          rlang::sym("t"), rlang::syms(c("x", "y"))
        ),
        names = name,
        tar_target(
          tar_name, val
        )
      )
    )
  })
  targets::tar_visnetwork()
})
```

![image](https://github.com/ropensci/tarchetypes/assets/13306020/2fe8787e-4aaa-48e3-bb53-e6d8d15530d1)

Is this by design? The `tar_eval()` works better here because we can specify the names of the targets in `values` too.

**Comment by wlandau** [comment, marked as answer]
`substitute(list(x), env = list(x = rlang::syms(list("y", "z"))))` and `substitute(list(x), env = list(x = quote(list(y, z))))` may look the same when you print them, but the returned objects are actually very different. The former has a non-language list object in place of `x`, whereas the latter is a better language object with true language objects at every level of the abstract syntax tree. `tar_eval()` is built on `substitute()`, so it works the same way. The following `_targets.R` file does what you want.

```r
# _targets.R
library(targets)
list(
  tar_target(x, 1),
  tar_target(y, 2),
  tarchetypes::tar_eval(
    tar_target(name, list(val)),
    values = tibble::tribble(
      ~name, ~val,
      quote(t), quote(list(x, y))
    )
  )
)
```

**Reply by psychelzh** [reply]
Thanks! Does it mean that I cannot create this type of dependency based on string input of "x" and "y" (these names are created programatically)?

**Reply by psychelzh** [reply]
Found it now, I should use `bquote()` to quasi-quote my expressions.

**Reply by psychelzh** [reply]
After a second thought, I found `bquote()` does not always help because I have many such names stored in a character vector. I cannot create such an object. For example: say I have two rows:

```r
tibble::tribble(
  ~name, ~val,
  "x", c("var_x1", "var_x2"),
  "y", c("var_y1", "var_y2"),
)
```

Each row depends on the two names (in `character` class) in `val` column, making `bquote()` hard to program with.

**Reply by psychelzh** [reply]
Several days later, I finally make it clear now. Initially, I had a misunderstanding about the abstract syntax tree. If symbols were created beforehand and nested into list as another variable, the abstract syntax tree was ruined. 

Now I abandoned the method of nesting those symbols beforehand and separate them into different columns. In this way, it even makes the code cleaner. Thanks again for your great package! 

(BTW, see https://github.com/psychelzh/g-factor/commit/d4649f463089955ee3ea11d85eb55531e4b03365 for my solution if anyone is interested.)
